include/caffe/datum_DB.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef CAFFE_DATUMDB_HPP
+#define CAFFE_DATUMDB_HPP
+
+#include <string>
+
+#include "leveldb/db.h"
+#include "lmdb.h"
+
+#include "caffe/common.hpp"
+#include "caffe/datum_DB_factory.hpp"
+#include "caffe/proto/caffe.pb.h"
+
+namespace caffe {
+
+class DatumDBCursor {
+ public:
+  explicit DatumDBCursor(const DatumDBParameter& param)
+    : param_(param) {}
+  ~DatumDBCursor() {
+    LOG(INFO) << "Closing DatumDBCursor on " << param_.source();
+  }
+
+  virtual bool Valid() = 0;
+  virtual void SeekToFirst() = 0;
+  virtual void Next() = 0;
+  virtual string key() = 0;
+  virtual Datum value() = 0;
+
+ protected:
+  DatumDBParameter param_;
+  DISABLE_COPY_AND_ASSIGN(DatumDBCursor);
+};
+
+class DatumDB {
+ public:
+  explicit DatumDB(const DatumDBParameter& param)
+    : param_(param),
+      is_opened_(new bool(false)) {}
+  virtual ~DatumDB() {
+    if (is_opened_.unique()) {
+      DatumDBRegistry::RemoveSource(param_.source());
+    }
+  }
+
+  virtual bool Get(const string& key, Datum* value) = 0;
+  virtual void Put(const string& key, const Datum& value) = 0;
+  virtual void Commit() = 0;
+  virtual DatumDBCursor* NewCursor() = 0;
+
+ protected:
+  virtual void Open() = 0;
+  virtual void Close() = 0;
+
+  DatumDBParameter param_;
+  shared_ptr<bool> is_opened_;
```

Comment:
Surely this trickery is no longer necessary, esp. given that copy is explicitly disabled on the next line?